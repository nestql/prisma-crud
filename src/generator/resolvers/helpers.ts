import { OptionalKind, MethodDeclarationStructure, Writers } from "ts-morph";

import { DmmfDocument } from "../dmmf/dmmf-document";
import { DMMF } from "../dmmf/types";

export function generateCrudResolverClassMethodDeclaration(
  action: DMMF.Action,
  mapping: DMMF.ModelMapping,
  dmmfDocument: DmmfDocument,
): OptionalKind<MethodDeclarationStructure> {
  return {
    name: action.name,
    isAsync: true,
    returnType: `Promise<${action.returnTSType}>`,
    decorators: [
      {
        name: `TypeGraphQL.${action.operation}`,
        arguments: [
          `_returns => ${action.typeGraphQLType}`,
          Writers.object({
            nullable: `${!action.method.isRequired}`,
          }),
        ],
      },
    ],
    parameters: [
      {
        name: "ctx",
        // TODO: import custom `ContextType`
        type: "any",
        decorators: [{ name: "TypeGraphQL.Context", arguments: [] }],
      },
      {
        name: "info",
        type: "GraphQLResolveInfo",
        decorators: [{ name: "TypeGraphQL.Info", arguments: [] }],
      },
      ...(!action.argsTypeName
        ? []
        : [
            {
              name: "args",
              type: action.argsTypeName,
              decorators: [{ name: "TypeGraphQL.Args", arguments: [] }],
            },
          ]),
    ],
    statements:
      action.kind === DMMF.ModelAction.aggregate
        ? [
            /* ts */ ` return getPrismaFromContext(ctx).${mapping.collectionName}.${action.prismaMethod}({
              ...args,
              ...transformFields(graphqlFields(info as any)),
            });`,
          ]
        : action.kind === DMMF.ModelAction.groupBy
        ? [
            /* ts */ ` const { _count, _avg, _sum, _min, _max } = transformFields(
              graphqlFields(info as any)
            );`,
            /* ts */ ` return getPrismaFromContext(ctx).${mapping.collectionName}.${action.prismaMethod}({
              ...args,
              ...Object.fromEntries(
                Object.entries({ _count, _avg, _sum, _min, _max }).filter(([_, v]) => v != null)
              ),
            });`,
          ]
        : [
            /* ts */ ` const { _count } = transformFields(
              graphqlFields(info as any)
            );
            return getPrismaFromContext(ctx).${mapping.collectionName}.${action.prismaMethod}({
              ...args,
              ...(_count && transformCountFieldIntoSelectRelationsCount(_count)),
            });`,
          ],
  };
}
