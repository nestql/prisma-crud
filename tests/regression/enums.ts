import { promises as fs } from "fs";

import generateArtifactsDirPath from "../helpers/artifacts-dir";
import { generateCodeFromSchema } from "../helpers/generate-code";
import createReadGeneratedFile, {
  ReadGeneratedFile,
} from "../helpers/read-file";

describe("enums", () => {
  let outputDirPath: string;
  let readGeneratedFile: ReadGeneratedFile;

  beforeEach(async () => {
    outputDirPath = generateArtifactsDirPath("regression-enums");
    await fs.mkdir(outputDirPath, { recursive: true });
    readGeneratedFile = createReadGeneratedFile(outputDirPath);
  });

  it("should properly generate code for normal enum", async () => {
    const schema = /* prisma */ `
      enum Color {
        RED
        GREEN
        BLUE
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const colorEnumTSFile = await readGeneratedFile("/enums/Color.ts");

    expect(colorEnumTSFile).toMatchSnapshot();
  });

  it("should properly generate code for enum with docs", async () => {
    const schema = /* prisma */ `
      /// Role enum doc
      enum Role {
        // User member comment
        USER
        ADMIN
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const roleEnumTSFile = await readGeneratedFile("/enums/Role.ts");
    const enumsIndexTSFile = await readGeneratedFile("/enums/index.ts");

    expect(roleEnumTSFile).toMatchSnapshot();
    expect(enumsIndexTSFile).toMatchSnapshot("enums index");
  });

  it("should properly generate standard prisma enums", async () => {
    const schema = /* prisma */ `
      model SampleModel {
        intIdField            Int     @id @default(autoincrement())
        stringField           String  @unique
        optionalStringField   String?
        intField              Int
        optionalIntField      Int?
        floatField            Float
        optionalFloatField    Float?
        booleanField          Boolean
        optionalBooleanField  Boolean?
        dateField             DateTime
        optionalDateField     DateTime?
        jsonField             Json
        optionalJsonField     Json?
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const sortOrderTSFile = await readGeneratedFile("/enums/SortOrder.ts");
    const queryModeTSFile = await readGeneratedFile("/enums/QueryMode.ts");
    const jsonNullValueFilterTSFile = await readGeneratedFile(
      "/enums/JsonNullValueFilter.ts",
    );
    const jsonNullValueInputTSFile = await readGeneratedFile(
      "/enums/JsonNullValueInput.ts",
    );
    const enumsIndexTSFile = await readGeneratedFile("/enums/index.ts");

    expect(sortOrderTSFile).toMatchSnapshot("SortOrder");
    expect(queryModeTSFile).toMatchSnapshot("QueryMode");
    expect(jsonNullValueFilterTSFile).toMatchSnapshot("JsonNullValueFilter");
    expect(jsonNullValueInputTSFile).toMatchSnapshot("JsonNullValueInput");
    expect(enumsIndexTSFile).toMatchSnapshot("enums index");
  });

  it("should properly generate model scalar field enum", async () => {
    const schema = /* prisma */ `
      model SampleModel {
        intIdField   Int     @id @default(autoincrement())
        stringField  String  @unique
        intField     Int
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const sampleModelScalarFieldEnumTSFile = await readGeneratedFile(
      "/enums/SampleModelScalarFieldEnum.ts",
    );
    const enumsIndexTSFile = await readGeneratedFile("/enums/index.ts");

    expect(sampleModelScalarFieldEnumTSFile).toMatchSnapshot(
      "SampleModelScalarFieldEnum",
    );
    expect(enumsIndexTSFile).toMatchSnapshot("enums index");
  });

  it("should properly generate model scalar enum when model is renamed", async () => {
    const schema = /* prisma */ `
      /// @@TypeGraphQL.type(name: "ExampleModel")
      model SampleModel {
        intIdField   Int     @id @default(autoincrement())
        stringField  String  @unique
        intField     Int
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const exampleModelScalarFieldEnumTSFile = await readGeneratedFile(
      "/enums/ExampleModelScalarFieldEnum.ts",
    );
    const enumsIndexTSFile = await readGeneratedFile("/enums/index.ts");

    expect(exampleModelScalarFieldEnumTSFile).toMatchSnapshot(
      "ExampleModelScalarFieldEnum",
    );
    expect(enumsIndexTSFile).toMatchSnapshot("enums index");
  });

  it("should properly generate model scalar enum when model field is renamed", async () => {
    const schema = /* prisma */ `
      model SampleModel {
        intIdField   Int     @id @default(autoincrement())
        /// @TypeGraphQL.field(name: "mappedFieldName")
        stringField  String  @unique
        intField     Int
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const sampleModelScalarFieldEnumTSFile = await readGeneratedFile(
      "/enums/SampleModelScalarFieldEnum.ts",
    );
    const enumsIndexTSFile = await readGeneratedFile("/enums/index.ts");

    expect(sampleModelScalarFieldEnumTSFile).toMatchSnapshot(
      "SampleModelScalarFieldEnum",
    );
    expect(enumsIndexTSFile).toMatchSnapshot("enums index");
  });

  describe("when `fullTextSearch` preview feature is enabled", () => {
    it("should properly generate enums for relevance", async () => {
      const schema = /* prisma */ `
        model SampleModel {
          intIdField   Int     @id @default(autoincrement())
          stringField  String  @unique
          intField     Int
        }
      `;

      await generateCodeFromSchema(schema, {
        outputDirPath,
        previewFeatures: ["fullTextSearch"],
      });
      const sampleModelOrderByRelevanceFieldEnumTSFile =
        await readGeneratedFile(
          "/enums/SampleModelOrderByRelevanceFieldEnum.ts",
        );

      expect(sampleModelOrderByRelevanceFieldEnumTSFile).toMatchSnapshot(
        "SampleModelOrderByRelevanceFieldEnum",
      );
    });

    describe("when model is renamed", () => {
      it("should properly generate enums for relevance", async () => {
        const schema = /* prisma */ `
          /// @@TypeGraphQL.type(name: "SampleRenamedModel")
          model SampleModel {
            intIdField   Int     @id @default(autoincrement())
            stringField  String  @unique
            intField     Int
          }
        `;

        await generateCodeFromSchema(schema, {
          outputDirPath,
          previewFeatures: ["fullTextSearch"],
        });
        const sampleRenamedModelOrderByRelevanceFieldEnumTSFile =
          await readGeneratedFile(
            "/enums/SampleRenamedModelOrderByRelevanceFieldEnum.ts",
          );

        expect(
          sampleRenamedModelOrderByRelevanceFieldEnumTSFile,
        ).toMatchSnapshot("SampleRenamedModelOrderByRelevanceFieldEnum");
      });
    });
  });

  describe("when `orderByNulls` preview feature is enabled", () => {
    it("should properly generate NullsOrder enum type", async () => {
      const schema = /* prisma */ `
        model FirstModel {
          idField             Int            @id @default(autoincrement())
          uniqueStringField   String         @unique
          optionalFloatField  Float?
          secondModelsField   SecondModel[]
        }
        model SecondModel {
          idField             Int          @id @default(autoincrement())
          uniqueStringField   String       @unique
          optionalFloatField  Float?
          firstModelFieldId   Int
          firstModelField     FirstModel   @relation(fields: [firstModelFieldId], references: [idField])
        }
      `;

      await generateCodeFromSchema(schema, {
        outputDirPath,
        previewFeatures: ["orderByNulls"],
      });
      const nullsOrderTSFile = await readGeneratedFile("/enums/NullsOrder.ts");
      const enumsIndexTSFile = await readGeneratedFile("/enums/index.ts");

      expect(nullsOrderTSFile).toMatchSnapshot("NullsOrder");
      expect(enumsIndexTSFile).toMatchSnapshot("enums index");
    });
  });
});
