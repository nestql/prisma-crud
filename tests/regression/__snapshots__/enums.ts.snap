// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enums should properly generate code for enum with docs 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

/** Role enum doc */
export enum Role {
  USER = \\"USER\\",
  ADMIN = \\"ADMIN\\"
}
TypeGraphQL.registerEnumType(Role, {
  name: \\"Role\\",
  description: \\"Role enum doc\\",
});
"
`;

exports[`enums should properly generate code for enum with docs: enums index 1`] = `
"export { Role } from \\"./Role\\";
export { TransactionIsolationLevel } from \\"./TransactionIsolationLevel\\";
"
`;

exports[`enums should properly generate code for normal enum 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

export enum Color {
  RED = \\"RED\\",
  GREEN = \\"GREEN\\",
  BLUE = \\"BLUE\\"
}
TypeGraphQL.registerEnumType(Color, {
  name: \\"Color\\",
  description: undefined,
});
"
`;

exports[`enums should properly generate model scalar enum when model field is renamed: SampleModelScalarFieldEnum 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

export enum SampleModelScalarFieldEnum {
  intIdField = \\"intIdField\\",
  mappedFieldName = \\"stringField\\",
  intField = \\"intField\\"
}
TypeGraphQL.registerEnumType(SampleModelScalarFieldEnum, {
  name: \\"SampleModelScalarFieldEnum\\",
  description: undefined,
});
"
`;

exports[`enums should properly generate model scalar enum when model field is renamed: enums index 1`] = `
"export { QueryMode } from \\"./QueryMode\\";
export { SampleModelScalarFieldEnum } from \\"./SampleModelScalarFieldEnum\\";
export { SortOrder } from \\"./SortOrder\\";
export { TransactionIsolationLevel } from \\"./TransactionIsolationLevel\\";
"
`;

exports[`enums should properly generate model scalar enum when model is renamed: ExampleModelScalarFieldEnum 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

export enum ExampleModelScalarFieldEnum {
  intIdField = \\"intIdField\\",
  stringField = \\"stringField\\",
  intField = \\"intField\\"
}
TypeGraphQL.registerEnumType(ExampleModelScalarFieldEnum, {
  name: \\"ExampleModelScalarFieldEnum\\",
  description: undefined,
});
"
`;

exports[`enums should properly generate model scalar enum when model is renamed: enums index 1`] = `
"export { ExampleModelScalarFieldEnum } from \\"./ExampleModelScalarFieldEnum\\";
export { QueryMode } from \\"./QueryMode\\";
export { SortOrder } from \\"./SortOrder\\";
export { TransactionIsolationLevel } from \\"./TransactionIsolationLevel\\";
"
`;

exports[`enums should properly generate model scalar field enum: SampleModelScalarFieldEnum 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

export enum SampleModelScalarFieldEnum {
  intIdField = \\"intIdField\\",
  stringField = \\"stringField\\",
  intField = \\"intField\\"
}
TypeGraphQL.registerEnumType(SampleModelScalarFieldEnum, {
  name: \\"SampleModelScalarFieldEnum\\",
  description: undefined,
});
"
`;

exports[`enums should properly generate model scalar field enum: enums index 1`] = `
"export { QueryMode } from \\"./QueryMode\\";
export { SampleModelScalarFieldEnum } from \\"./SampleModelScalarFieldEnum\\";
export { SortOrder } from \\"./SortOrder\\";
export { TransactionIsolationLevel } from \\"./TransactionIsolationLevel\\";
"
`;

exports[`enums should properly generate standard prisma enums: JsonNullValueFilter 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

export enum JsonNullValueFilter {
  DbNull = \\"DbNull\\",
  JsonNull = \\"JsonNull\\",
  AnyNull = \\"AnyNull\\"
}
TypeGraphQL.registerEnumType(JsonNullValueFilter, {
  name: \\"JsonNullValueFilter\\",
  description: undefined,
});
"
`;

exports[`enums should properly generate standard prisma enums: JsonNullValueInput 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

export enum JsonNullValueInput {
  JsonNull = \\"JsonNull\\"
}
TypeGraphQL.registerEnumType(JsonNullValueInput, {
  name: \\"JsonNullValueInput\\",
  description: undefined,
});
"
`;

exports[`enums should properly generate standard prisma enums: QueryMode 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

export enum QueryMode {
  \\"default\\" = \\"default\\",
  insensitive = \\"insensitive\\"
}
TypeGraphQL.registerEnumType(QueryMode, {
  name: \\"QueryMode\\",
  description: undefined,
});
"
`;

exports[`enums should properly generate standard prisma enums: SortOrder 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

export enum SortOrder {
  asc = \\"asc\\",
  desc = \\"desc\\"
}
TypeGraphQL.registerEnumType(SortOrder, {
  name: \\"SortOrder\\",
  description: undefined,
});
"
`;

exports[`enums should properly generate standard prisma enums: enums index 1`] = `
"export { JsonNullValueFilter } from \\"./JsonNullValueFilter\\";
export { JsonNullValueInput } from \\"./JsonNullValueInput\\";
export { NullableJsonNullValueInput } from \\"./NullableJsonNullValueInput\\";
export { QueryMode } from \\"./QueryMode\\";
export { SampleModelScalarFieldEnum } from \\"./SampleModelScalarFieldEnum\\";
export { SortOrder } from \\"./SortOrder\\";
export { TransactionIsolationLevel } from \\"./TransactionIsolationLevel\\";
"
`;

exports[`enums when \`fullTextSearch\` preview feature is enabled should properly generate enums for relevance: SampleModelOrderByRelevanceFieldEnum 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

export enum SampleModelOrderByRelevanceFieldEnum {
  stringField = \\"stringField\\"
}
TypeGraphQL.registerEnumType(SampleModelOrderByRelevanceFieldEnum, {
  name: \\"SampleModelOrderByRelevanceFieldEnum\\",
  description: undefined,
});
"
`;

exports[`enums when \`fullTextSearch\` preview feature is enabled when model is renamed should properly generate enums for relevance: SampleRenamedModelOrderByRelevanceFieldEnum 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

export enum SampleRenamedModelOrderByRelevanceFieldEnum {
  stringField = \\"stringField\\"
}
TypeGraphQL.registerEnumType(SampleRenamedModelOrderByRelevanceFieldEnum, {
  name: \\"SampleRenamedModelOrderByRelevanceFieldEnum\\",
  description: undefined,
});
"
`;

exports[`enums when \`orderByNulls\` preview feature is enabled should properly generate NullsOrder enum type: NullsOrder 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";

export enum NullsOrder {
  first = \\"first\\",
  last = \\"last\\"
}
TypeGraphQL.registerEnumType(NullsOrder, {
  name: \\"NullsOrder\\",
  description: undefined,
});
"
`;

exports[`enums when \`orderByNulls\` preview feature is enabled should properly generate NullsOrder enum type: enums index 1`] = `
"export { FirstModelScalarFieldEnum } from \\"./FirstModelScalarFieldEnum\\";
export { NullsOrder } from \\"./NullsOrder\\";
export { QueryMode } from \\"./QueryMode\\";
export { SecondModelScalarFieldEnum } from \\"./SecondModelScalarFieldEnum\\";
export { SortOrder } from \\"./SortOrder\\";
export { TransactionIsolationLevel } from \\"./TransactionIsolationLevel\\";
"
`;
